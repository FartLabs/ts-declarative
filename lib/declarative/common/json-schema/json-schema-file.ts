import { Project } from "ts-morph";
import type { Class } from "#/lib/declarative/declarative.ts";
import { declarativeTsMorph } from "#/lib/declarative/common/ts-morph/ts-morph.ts";
import { createDecoratorFactory } from "#/lib/declarative/decorator.ts";
import { declarativeJSONSchema } from "./json-schema.ts";

export async function jsonSchemaDecoratorFactoryOfFile(
  specifier: string | URL,
): Promise<() => (target: Class) => Class> {
  const project = new Project({ useInMemoryFileSystem: true });
  project.createSourceFile(
    specifier.toString(),
    await Deno.readTextFile(new URL(specifier)),
  );

  return createDecoratorFactory({
    initialize: () => {
      const sourceFile = project.getSourceFileOrThrow(specifier.toString());
      return [declarativeTsMorph(sourceFile), declarativeJSONSchema()];
    },
  });
}
