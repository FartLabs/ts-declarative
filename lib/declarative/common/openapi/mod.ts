export * from "./paths-object.ts";
export * from "./specification.ts";
export * from "./server.ts";
