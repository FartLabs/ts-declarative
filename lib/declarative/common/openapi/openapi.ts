import type { Route } from "@std/http/unstable-route";
import type { OpenAPIV3_1 } from "openapi-types";
import type { Class, Declarative } from "#/lib/declarative/declarative.ts";
import { getPrototypeValue } from "#/lib/declarative/declarative.ts";
import { createDecoratorFactory } from "#/lib/declarative/decorator.ts";

/**
 * specificationOf returns the OpenAPI specification of the class.
 */
export function specificationOf<TClass extends Class>(
  target: TClass,
): OpenAPIV3_1.Document | undefined {
  return getPrototypeValue<ValueOpenAPI>(target)?.specification;
}

/**
 * routesOf returns the HTTP routes of the OpenAPI class.
 */
export function routesOf<TClass extends Class>(
  target: TClass,
): Route[] | undefined {
  return getPrototypeValue<ValueOpenAPI>(target)?.routes;
}

export function openapiDecoratorFactory(): () => (target: Class) => Class {
  return createDecoratorFactory({
    initialize: (value?: ValueOpenAPI) => {
      return [declarativeOpenAPI(value)];
    },
  });
}

export function declarativeOpenAPI<TValue extends ValueOpenAPI>(
  value0?: TValue,
): Declarative<TValue> {
  return (value1) => {
    return Object.assign({}, value0, value1);
  };
}

/**
 * ValueOpenAPI is the value associated with an OpenAPI specification.
 */
export interface ValueOpenAPI {
  /**
   * specification is the OpenAPI specification.
   */
  specification?: OpenAPIV3_1.Document;

  /**
   * routes are the HTTP routes of the OpenAPI specification.
   */
  routes?: Route[];

  /**
   * resources are the resources registered in the OpenAPI specification.
   */
  resources?: Class[];
}
