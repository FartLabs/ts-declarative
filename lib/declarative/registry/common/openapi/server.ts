// import type { Route } from "@std/http/unstable-route";
