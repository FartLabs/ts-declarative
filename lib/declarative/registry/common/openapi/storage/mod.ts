export * from "./storage.ts";
export * from "./memory.ts";
export * from "./deno-kv.ts";
