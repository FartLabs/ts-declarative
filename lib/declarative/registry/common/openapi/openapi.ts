import type { Route } from "@std/http/unstable-route";
import { slugify } from "@std/text/unstable-slugify";
import type { OpenAPIV3_1 } from "openapi-types";
import type { Class } from "#/lib/declarative/declarative.ts";
import type { Registry } from "#/lib/declarative/registry/registry.ts";
import { jsonSchemaOf } from "#/lib/declarative/common/json-schema/json-schema.ts";

/**
 * OpenAPIServer is a registry for OpenAPI resources that builds an
 * OpenAPI specification.
 */
export class OpenAPIServer implements Registry {
  public specification: OpenAPIV3_1.Document;
  public routes: Route[] = []; // TODO: Append routes on register.

  public constructor(specification?: Partial<OpenAPIV3_1.Document>) {
    this.specification = {
      ...specification,
      openapi: specification?.openapi ?? "3.1.0",
      info: specification?.info ?? { title: "Example API", version: "1.0.0" },
      paths: specification?.paths ?? {},
    };
  }

  public register(target: Class, options?: OpenAPIResourceOptions): void {
    const jsonSchema = jsonSchemaOf(target);
    if (jsonSchema === undefined) {
      throw new Error("Target must have a JSON schema");
    }

    const { path, endpoints } = openapiResourceOptions(target, options);
    if (endpoints.create) {
      const operationPath = operationPathOf(path, "create");
      {
        if (this.hasOperation(operationPath, methodOf("create"))) {
          throw new Error(
            `Path ${operationPath} already has a create operation.`,
          );
        }

        this.setCreateOperation(operationPath, jsonSchema);
      }

      if (endpoints.read) {
        const operationPath = operationPathOf(path, "read");
        if (this.hasOperation(operationPath, methodOf("read"))) {
          throw new Error(
            `Path ${operationPath} already has a read operation.`,
          );
        }

        this.setReadOperation(operationPath, jsonSchema);
      }

      if (endpoints.update) {
        const operationPath = operationPathOf(path, "update");
        if (this.hasOperation(operationPath, methodOf("update"))) {
          throw new Error(
            `Path ${operationPath} already has an update operation.`,
          );
        }

        this.setUpdateOperation(operationPath, jsonSchema);
      }

      if (endpoints.delete) {
        const operationPath = operationPathOf(path, "delete");
        if (this.hasOperation(operationPath, methodOf("delete"))) {
          throw new Error(
            `Path ${operationPath} already has a delete operation.`,
          );
        }

        this.setDeleteOperation(operationPath);
      }
    }
  }

  public hasOperation(path: string, method: string): boolean {
    return Object.hasOwn(this.specification.paths?.[path] ?? {}, method);
  }

  // deno-lint-ignore no-explicit-any
  public setOperation(path: string, method: string, data: any): void {
    this.specification.paths ??= {};
    this.specification.paths[path] ??= {};
    this.specification.paths[path][method as ReturnType<typeof methodOf>] =
      data;
  }

  public setCreateOperation<T>(path: string, jsonSchema: T): void {
    this.setOperation(path, methodOf("create"), {
      requestBody: {
        content: {
          // TODO: Use $ref instead of schema directly.
          "application/json": { schema: jsonSchema },
        },
      },
      responses: { "200": { description: "Created" } },
    });
  }

  public setReadOperation<T>(path: string, jsonSchema: T): void {
    this.setOperation(path, methodOf("read"), {
      parameters: [idPathParameter],
      responses: {
        "200": {
          description: "OK",
          content: {
            // TODO: Use $ref instead of schema directly.
            "application/json": { schema: jsonSchema },
          },
        },
      },
    });
  }

  public setUpdateOperation<T>(path: string, jsonSchema: T): void {
    this.setOperation(path, methodOf("update"), {
      parameters: [idPathParameter],
      requestBody: {
        content: { "application/json": { schema: jsonSchema } },
      },
      responses: { "200": { description: "Updated" } },
    });
  }

  public setDeleteOperation(path: string): void {
    this.setOperation(path, methodOf("delete"), {
      parameters: [idPathParameter],
      responses: {
        "200": { description: "Deleted" },
      },
    });
  }
}

const idPathParameter: OpenAPIV3_1.ParameterObject = {
  name: "id",
  in: "path",
  required: true,
  schema: { type: "string" },
};

/**
 * openapiResourceOptions creates options for registering an opinionated
 * OpenAPI resource.
 */
export function openapiResourceOptions(
  target: Class,
  options?: OpenAPIResourceOptions,
): Required<OpenAPIResourceOptions> {
  const { path, endpoints } = options ?? {};
  return {
    path: path ?? `/${slugify(target.name)}`,
    endpoints: { ...defaultOpenAPIResourceEndpointKinds(), ...endpoints },
  };
}

/**
 * OpenAPIResourceOptions are options for registering a resource and its
 * endpoints on an OpenAPI specification.
 */
export interface OpenAPIResourceOptions {
  path?: string;
  endpoints?: Record<OpenAPIResourceEndpointKind, boolean>;
}

export function methodOf(
  endpointKind: OpenAPIResourceEndpointKind,
): "get" | "post" | "delete" {
  switch (endpointKind) {
    case "create": {
      return "post";
    }

    case "read": {
      return "get";
    }

    case "update": {
      return "post";
    }

    case "delete": {
      return "delete";
    }

    default: {
      throw new Error(`Unknown endpoint kind: ${endpointKind}`);
    }
  }
}

export function operationPathOf(
  prefix: string,
  endpointKind: OpenAPIResourceEndpointKind,
): string {
  switch (endpointKind) {
    case "create": {
      return prefix;
    }

    case "read":
    case "update":
    case "delete": {
      return `${prefix}/{id}`;
    }
  }
}

export function defaultOpenAPIResourceEndpointKinds(): Record<
  OpenAPIResourceEndpointKind,
  boolean
> {
  return Object.fromEntries(
    openapiResourceEndpointKinds.map((endpoint) => {
      return [endpoint, true];
    }),
  ) as Record<OpenAPIResourceEndpointKind, boolean>;
}

export type OpenAPIResourceEndpointKind =
  (typeof openapiResourceEndpointKinds)[number];

export const openapiResourceEndpointKinds: [
  "create",
  "read",
  "update",
  "delete",
] = [
  "create",
  "read",
  "update",
  "delete",
  // "batchCreate", "batchRead", "batchUpdate", "batchDelete",
];
