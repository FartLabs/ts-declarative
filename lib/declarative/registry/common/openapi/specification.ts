// import { slugify } from "@std/text/unstable-slugify";
import type { Class } from "#/lib/declarative/declarative.ts";
import { getPrototypeID } from "#/lib/declarative/declarative.ts";
import type { Registry } from "#/lib/declarative/registry/registry.ts";
import { jsonSchemaOf } from "#/lib/declarative/common/json-schema/json-schema.ts";

/**
 * OpenAPISpecification is a registry for OpenAPI resources that builds an
 * OpenAPI specification.
 */
export class OpenAPISpecification implements Registry<OpenAPIResourceOptions> {
  // deno-lint-ignore no-explicit-any
  public constructor(public specification: any) {}

  public register(target: Class, _options?: OpenAPIResourceOptions): void {
    const id = getPrototypeID(target);
    if (id === undefined) {
      throw new Error("Target must be a class");
    }

    const jsonSchema = jsonSchemaOf(target);
    if (jsonSchema === undefined) {
      throw new Error("Target must have a JSON schema");
    }

    // TODO: Define CRUD endpoints on base path.
    // TODO: Generate client library with oazapfts.
    // const path = options?.path ?? slugify(id);

    throw new Error("Not implemented.");
  }
}

/**
 * OpenAPIResourceOptions are options for registering an OpenAPI resource.
 */
export interface OpenAPIResourceOptions {
  path?: string;
}
