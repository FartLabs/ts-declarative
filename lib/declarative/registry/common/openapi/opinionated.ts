import type { Class } from "#/lib/declarative/declarative.ts";
import { MuxRegistry } from "#/lib/declarative/registry/mux.ts";
import { OpenAPISpecification } from "#/lib/declarative/registry/common/openapi/specification.ts";

/**
 * opinionatedOpenAPI creates an opinionated OpenAPI registry.
 */
export function opinionatedOpenAPI() {
  return new MuxRegistry([new OpenAPISpecification({})]);
}

/**
 * opinionatedResourceOptions creates options for registering an opinionated
 * OpenAPI resource.
 */
export function opinionatedResourceOptions(
  target: Class,
  options?: OpinionatedResourceOptions,
): OpinionatedResourceOptions {
  const { path, endpoints } = options ?? {};
  return {
    path: path ?? target.name,
    endpoints: endpoints ?? defaultOpinionatedResourceEndpointKinds(),
  };
}

export interface OpinionatedResourceOptions {
  path?: string;
  endpoints?: Record<OpinionatedResourceEndpointKind, boolean>;
}

export type OpinionatedResourceEndpointKind =
  (typeof opinionatedResourceEndpointKinds)[number];

const opinionatedResourceEndpointKinds = [
  "create",
  "read",
  "update",
  "delete",
  // "batchCreate", "batchRead", "batchUpdate", "batchDelete",
] as const satisfies string[];

function defaultOpinionatedResourceEndpointKinds(): Record<
  OpinionatedResourceEndpointKind,
  boolean
> {
  return Object.fromEntries(
    opinionatedResourceEndpointKinds.map((endpoint) => {
      return [endpoint, true];
    }),
  ) as Record<OpinionatedResourceEndpointKind, boolean>;
}
