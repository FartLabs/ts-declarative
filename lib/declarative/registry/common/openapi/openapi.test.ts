import { assertSnapshot } from "@std/testing/snapshot";
import { jsonSchemaDecoratorFactoryOfFile } from "#/lib/declarative/common/json-schema/json-schema-file.ts";
import { OpenAPIServer } from "./openapi.ts";

const jsonSchema = await jsonSchemaDecoratorFactoryOfFile(import.meta.url);

@jsonSchema()
class Example {
  public constructor(public name: string) {}
}

Deno.test("OpenAPIServer registers class", async (t) => {
  const server = new OpenAPIServer();
  server.register(Example);

  await assertSnapshot(t, JSON.stringify(server.specification, null, 2));
});
