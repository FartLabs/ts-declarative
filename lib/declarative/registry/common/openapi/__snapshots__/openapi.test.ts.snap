export const snapshot = {};

snapshot[`OpenAPIServer registers class 1`] = `
'{
  "openapi": "3.1.0",
  "info": {
    "title": "Example API",
    "version": "1.0.0"
  },
  "paths": {
    "/example": {
      "post": {
        "requestBody": {
          "content": {
            "application/json": {
              "schema": {
                "type": "object",
                "properties": {
                  "name": {
                    "type": "string"
                  }
                },
                "required": [
                  "name"
                ]
              }
            }
          }
        },
        "responses": {
          "200": {
            "description": "Created"
          }
        }
      }
    },
    "/example/{id}": {
      "get": {
        "parameters": [
          {
            "name": "id",
            "in": "path",
            "required": true,
            "schema": {
              "type": "string"
            }
          }
        ],
        "responses": {
          "200": {
            "description": "OK",
            "content": {
              "application/json": {
                "schema": {
                  "type": "object",
                  "properties": {
                    "name": {
                      "type": "string"
                    }
                  },
                  "required": [
                    "name"
                  ]
                }
              }
            }
          }
        }
      },
      "post": {
        "parameters": [
          {
            "name": "id",
            "in": "path",
            "required": true,
            "schema": {
              "type": "string"
            }
          }
        ],
        "requestBody": {
          "content": {
            "application/json": {
              "schema": {
                "type": "object",
                "properties": {
                  "name": {
                    "type": "string"
                  }
                },
                "required": [
                  "name"
                ]
              }
            }
          }
        },
        "responses": {
          "200": {
            "description": "Updated"
          }
        }
      },
      "delete": {
        "parameters": [
          {
            "name": "id",
            "in": "path",
            "required": true,
            "schema": {
              "type": "string"
            }
          }
        ],
        "responses": {
          "200": {
            "description": "Deleted"
          }
        }
      }
    }
  }
}'
`;
